worker/uniter/runner/jujuc/credential-get_test.go

```diff
@@ -0,0 +1,37 @@
+package jujuc_test
+
+import (
+	"github.com/juju/cmd"
+	"github.com/juju/cmd/cmdtesting"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/worker/uniter/runner/jujuc"
+)
+
+type CredentialGetSuite struct {
+	ContextSuite
+}
+
+var _ = gc.Suite(&CredentialGetSuite{})
+
+// [TODO](externalreality): Many jujuc commands can be run through a processor
+// much like the one below. This sort of thing should not have to be written
+// more than once except for in special cases. A structure containing all of the
+// relevant jujuc commands along with their supported format options would cut
+// down on a great deal of test fluff. The juju/cmd test are a good example of how
+// this might be done.
+func runCredentialGetCommand(s *CredentialGetSuite, c *gc.C, args []string) (*cmd.Context, int) {
+	hctx := s.GetHookContext(c, -1, "")
+	com, err := jujuc.NewCommand(hctx, cmdString("credential-get"))
+	c.Assert(err, jc.ErrorIsNil)
+	ctx := cmdtesting.Context(c)
+	code := cmd.Main(com, ctx, args)
+	return ctx, code
+}
+
+func (s *CredentialGetSuite) TestCommandRun(c *gc.C) {
+	_, exitCode := runCredentialGetCommand(s, c, []string{})
+	exitSuccess := 0
+	c.Assert(exitCode, gc.Equals, exitSuccess)
+}
```

Comment:
do we have any sort of test that the output of running "credential-get" is a properly formatted (can be parsed as yaml/json) and would contain the fields that you would care about?

That seems more like the sort of thing that we would concretely care about, as it is the promises that we are giving to charms "you can run credential-get and it will have these fields, and you can ask for it in --format=yaml or --format=json".

There are things like
"did we actually call AddFlags...DefaultFormatters"
that doesn't seem to be covered here.
We could probably seed the mock with a set of testable values. I'd probably seed it with aws-like credential (if only because that's clearly where we'll get the most usage at first).
Something that tests you can parse out the AUTH_URL, secret-id and auth-id (I forget the exact names).

Reply:
also, tests that when something *isn't* authorized, the charm gets a clear signal (maybe its just the exit code?) but some sort of clear indication that we promise to preserve for charms so that they can respond to that failure and know they're handling it correctly.

Reply:
(maybe some of that is better tested with an acceptance test running our own credential-consuming dummy charm. something that would help us as developers know that we can make a release without worrying that we broke this feature.)

Reply:
No we don't have any tests for parsing. Yet, a good way to do that my be to store the "formaters" on the command structure itself. This way we can generically iterate for matters and test for supplied keys. As far as testing for a defined set of keys there are two ways; one would be to double book keep a defined set of keys (those fields that are specified as necessary  by struct tags and those listed in the test), another would be to test the consumption of the output. I am partial to the latter which states "these keys are necessary because for this use case."

Reply:
"did we actually call AddFlags...DefaultFormatters" is definitely pragmatic and I am not above the approach. Yet, I always disk liked the technique since it definitely is a test of implementation" rather than behavior. As mentioned above I wouldn't mind a generic serialize/serialize test that loops through formats.